cmd/api/book/service.go

```diff
@@ -0,0 +1,147 @@
+package book
+
+import (
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+type ServiceAPI interface {
+	ArchiveBook(id uuid.UUID) (Book, error)
+	CreateBook(bookEntry Book) (Book, error)
+	GetBook(id uuid.UUID) (Book, error)
+	ListBooks(query url.Values) (PagedBooks, error)
+	UpdateBook(bookEntry Book, id uuid.UUID) (Book, error)
+}
+
+type Repository interface {
+	SetBookArchiveStatus(id uuid.UUID, archived bool) (Book, error)
+	CountRows(name string, minPrice32, maxPrice32 float32, archived bool) (int, error)
+	CreateBook(bookEntry Book) (Book, error)
+	GetBookByID(id uuid.UUID) (Book, error)
+	ListBooks(name string, minPrice32, maxPrice32 float32, sortBy, sortDirection string, archived bool, page, pageSize int) ([]Book, error)
+	UpdateBook(bookEntry Book) (Book, error)
+}
+
+type Service struct {
+	repo Repository
+}
+
+func NewService(repo Repository) *Service {
+	return &Service{repo: repo}
+}
+
+func (s *Service) ArchiveBook(id uuid.UUID) (Book, error) {
+	archived := true
+	return s.repo.SetBookArchiveStatus(id, archived)
+}
+
+func (s *Service) CreateBook(bookEntry Book) (Book, error) {
+	bookEntry.ID = uuid.New()                                      //Atribute an ID to the entry
+	bookEntry.CreatedAt = time.Now().UTC().Round(time.Millisecond) //Atribute creating and updating time to the new entry. UpdateAt can change later.
+	bookEntry.UpdatedAt = bookEntry.CreatedAt
+	return s.repo.CreateBook(bookEntry)
+}
+
+func (s *Service) GetBook(id uuid.UUID) (Book, error) {
+	return s.repo.GetBookByID(id)
+}
+
+type PagedBooks struct {
+	PageCurrent int    `json:"page_current"`
+	PageTotal   int    `json:"page_total"`
+	PageSize    int    `json:"page_size"`
+	ItemsTotal  int    `json:"items_total"`
+	Results     []Book `json:"results"`
+}
+
+func (s *Service) ListBooks(query url.Values) (PagedBooks, error) {
```

Comment:
You shouldn't be dealing with `url.Values` here. This is specific to the HTTP layer.
What you can do to pass the search parameters to the `ListBooks` method is create a `request` type.
```go
type ListBooksRequest struct {
    Name string
    MinPrice float
//...
}

Reply:
Then you would extract the values from the `http.Request` in the HTTP layer.

Reply:
I've tried to pass a variable of type http.ListBooksRequest to service.go, but then it felt in a cycle import error. So I had to pass the list of parameters extensively. It worked, at least.

Reply:
In the middle of the night I realized that I could create the struct type at service.go side!